rfd/0126-backend-migrations.md

```diff
@@ -0,0 +1,258 @@
+---
+authors: Tim Ross (tim.ross@goteleport.com)
+state: draft
+---
+
+# RFD 126 - Backend Migrations
+
+## Required Approvers
+
+- Engineering @zmb3 && (@fspmarshall || @espadollini )
+- Product: ( @klizhentas || @russjones )
+
+## What
+
+A more reliable and robust mechanism to perform backend migrations.
+
+## Why
+
+The upgrade process requires scaling Auth down to a single instance to ensure that migrations are only performed once as well as preventing a new version and an old version from operating on the same keys with different schema. This is cumbersome and makes the upgrade process a pain point for cluster admins. Scaling Auth down and back up also results in an [uneven load](https://github.com/gravitational/teleport/issues/7029) which can cause connectivity issues and backend latency that can result in a thundering herd.
+
+Migrations to date have existed as a function that is called during Auth initialization. The migration function is left as is for a release or two and is then deleted. This makes tracking when and which migrations have been applied difficult and prevents skipping any major versions when upgrading since there are no guarantees that the latest version will include all the migrations between versions.
+
+## Details
+
+### When is a Migration required?
+
+Migrations are not needed for every change to a backend resource. Adding a new field to the resource should be a backward compatible operation by nature of Protocol Buffers and does not require an actual migration. Any major changes to a resource that alter its shape, change its encoding, or moving a single resource into several other resources are candidates for a migration.
+
+### Persistent Migrations
+
+All new migrations MUST exist in perpetuity to allow for the correct migrations to be applied regardless of which version of Teleport is being upgraded from and to. Migrations MUST be numbered and applied in sequence. The order of migrations MUST not change and a migration MUST not be altered once it has been included in a release.
```

Comment:
Is being able to upgrade from any version to any newer version an explicit design goal or a nice to have? My understanding is it's the latter.

(To be clear, I really want this feature, but making it clear what the real constraints here will help in evaluating the solution)

Reply:
It is not a requirement, Zac. I don't think keeping all migration in perpetuity makes sense from the costs perspective. We only promise one major forward version upgrade. Although, this will be more relevant for cloud CI/CD, there will be no strict version.

Reply:
If we decide to limit migrations, we should have the auth server refuse to start if there's been (completed) migrations that it doesn't know about, or if some older migrations are missing and no longer supported; that way, users are prohibited from shooting themselves in the foot by skipping ahead in versions too much, whatever we decide "too much" is.